046-analyse-broker-metrics-before-rolling-kafka.md

```diff
@@ -0,0 +1,63 @@
+
+1. Add a web server to KafkaAgent that exposes an endpoint which returns a value of metric
+    - Configure Jetty server to only accept SSL connections on a port 8443.
+    - Expose an endpoint `/broker-state`.
+    - Add a servlet to handle a `GET` request and return an integer value of `BrokerState` metric.
+    - It will reuse broker's own certificate, which will already be in the container.
+2. Update KafkaRoller to query the broker state from the KafkaAgent API
+    - Currently KafkaRoller obtains headless service for a given pod and uses it to boostrap Kafka AdminClient. It can use the same mechanism to get the address for the KafkaAgent endpoint.
+    - Instead of checking broker's readiness through `isReady()` function, utilise the new restful endpoint to determine if the broker is ready.
+    - After restarting a broker if it does not become ready within the operational timeout period, check whether the broker is performing log recovery by querying the new endpoint:
+        - If log recovery is in progress, do not restart the broker.
+        - Otherwise, force restart the broker which is the current behabviour.
+    - Add a new configurable timeout for log recovery that is 60 minutes by default.
+    - If the log recovery does not complete within the timeout period, then fail the reconciliation and report an error for the human operator to investigate further. 
+
+### For future consideration:
+- The operator might need access to more broker metrics. We can easily implement additional endpoints to expose them.
+- The readiness check could also make use of the KafkaAgent endpoint to get the broker state instead of checking for a file on disk.
+- The liveness and readiness checks potentially will need to access non yammer metrics for KRAFT, which could be exposed in the same manner, but would need a new mechanism to collect them.
+
+## Affected/not affected projects
+
+* strimzi/kafka-cluster-operator
+
+## Compatibility
+
+When the operator gets updated with the KafkaRoller change, it will expect the KafkaAgent endpoint to be there. However, KafkaAgent would only get updated after the operator.
+
+1. Roll out the KafkaAgent changes for the new API.
+2. Roll out the KafkaRoller changes and gated behind a feature flag. By default, KafkaRoller will still have the current behaviour and will not use the KafkaAgent API. 
+3. User can enable KafkaRoller to use KafkaAgent API using the feature flag. The user has to make sure that they are running the version that has the KafkaAgent API before they enable the feature. 
+4. We can mark the feature flag deprecetad immediately. 
+5. In the next major release, we can remove the feature flag and KafkaRoller can always use the KafkaAgent API.
+
+## Rejected alternatives
+- Why not use VertX? Strimzi is moving away from using VertX. Kafka already uses Jetty which is convenient for KafkaAgent, as there is no need to introduce a new dependency.
```

Comment:
This is not really about Strimzi's use of Vert.x. Vert.x is used in operators only. It is not part of the Kafka images anyway. So regardless of how long we keep using it, Jetty which is already part of Kafka is a more logical choice.

Reply:
I agree, I think that we could avoid mentioning Vert.x here. The Jetty choice is the obvious one because it's already there.